### PR TITLE
feat: Store cache information for each sync flow in sam sync context

### PR DESCRIPTION
**Pull request by hnnasit**
#### Which issue(s) does this change fix?
N/A

#### Why is this change necessary?
Caching sync flow hashes can speed up the sync when used for comparing locally. This will also prevent throttling issues when the resource APIs are called for similar resources.

#### How does it address the issue?
Store the resource payload hash of each sync flow which can be used for calling resource apis only on the resources with different hashes preventing throttling issues.

#### What side effects does this change have?
None

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [x] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write/update unit tests
- [ ] Write/update integration tests
- [ ] Write/update functional tests if needed
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

**Review by mndeveci — APPROVED** [review]

File: tests/unit/commands/sync/test_sync_context.py [hunk 1]
```diff
@@ -7,31 +7,99 @@
...
+            (
+                False,
+                {
+                    "MockResourceId": ResourceSyncState("mock-hash", MOCK_TIME),
+                    "Parent/Child/MockResourceId": ResourceSyncState("nested-mock-hash", MOCK_TIME),
+                },
+            ),
+        ]
+    )
+    def test_sync_state(self, resource_sync_states, dependency_layer):
```
Comment (lucashuy) [inline tests/unit/commands/sync/test_sync_context.py:41]: Just double checking, are the function arguments reversed here?

**Review by lucashuy — COMMENTED** [review]
Implementation details look good to me, just had a comment about the tests

File: tests/unit/commands/sync/test_sync_context.py [hunk 2]
```diff
@@ -7,31 +7,99 @@
...
+        self.assertEqual(sync_state.dependency_layer, dependency_layer)
+        self.assertEqual(sync_state.resource_sync_states, resource_sync_states)
+
+    @parameterized.expand(
+        [
+            (True, "MockResourceId", "mock-hash"),
+            (False, "Parent/Child/MockResourceId", "mock-nested-hash"),
+        ]
+    )
+    def test_sync_state_update_resource_sync_state(self, resource_hash, resource_id, dependency_layer):
```
Comment (lucashuy) [inline tests/unit/commands/sync/test_sync_context.py:52]: (didn't realise I never pressed the add comment button) also here

File: tests/unit/commands/sync/test_sync_context.py [hunk 3]
```diff
@@ -7,31 +7,99 @@
...
+        self.assertEqual(sync_state.resource_sync_states[resource_id].hash_value, resource_hash)
+
+
+class TestResourceSyncState(TestCase):
+    @parameterized.expand(
+        [
+            ("mockhash", MOCK_TIME),
+        ]
+    )
+    def test_sync_state(self, sync_time, hash_str):
```
Comment (lucashuy) [inline tests/unit/commands/sync/test_sync_context.py:67]: (didn't realise I never pressed the add comment button) and here

File: tests/unit/commands/sync/test_sync_context.py (hunk 1 above)
Comment (hnnasit) [reply tests/unit/commands/sync/test_sync_context.py:41]: Good catch! weird that the tests didn't fail when the wrong type args were provided.

**Review by lucashuy — APPROVED** [review]
Thanks for making the changes!

**Pull request merged**